nemo/collections/nlp/models/qa_model.py

```diff
@@ -99,20 +102,72 @@ def validation_step(self, batch, batch_idx):
+
+        unique_ids = torch.cat([x['eval_tensors']['unique_ids'] for x in outputs])
+        start_logits = torch.cat([x['eval_tensors']['start_logits'] for x in outputs])
+        end_logits = torch.cat([x['eval_tensors']['end_logits'] for x in outputs])
+
+        all_unique_ids = []
+        all_start_logits = []
+        all_end_logits = []
+        if torch.distributed.is_initialized():
+            world_size = torch.distributed.get_world_size()
+            for ind in range(world_size):
+                all_unique_ids.append(torch.empty_like(unique_ids))
+                all_start_logits.append(torch.empty_like(start_logits))
+                all_end_logits.append(torch.empty_like(end_logits))
+            torch.distributed.all_gather(all_unique_ids, unique_ids)
+            torch.distributed.all_gather(all_start_logits, start_logits)
+            torch.distributed.all_gather(all_end_logits, end_logits)
+        else:
+            all_unique_ids.append(unique_ids)
+            all_start_logits.append(start_logits)
+            all_end_logits.append(end_logits)
+
+        exact_match, f1, all_predictions, all_nbest = -1, -1, [], []
+        if not torch.distributed.is_initialized() or torch.distributed.get_rank() == 0:
+
+            unique_ids = []
+            start_logits = []
+            end_logits = []
+            for u in all_unique_ids:
+                unique_ids.extend(u.cpu().numpy().tolist())
+            for u in all_start_logits:
+                start_logits.extend(u.cpu().numpy().tolist())
+            for u in all_end_logits:
+                end_logits.extend(u.cpu().numpy().tolist())
+
+            exact_match, f1, all_predictions, all_nbest = self.validation_dataset.evaluate(
+                unique_ids=unique_ids,
+                start_logits=start_logits,
+                end_logits=end_logits,
+                n_best_size=self.validation_config.n_best_size,
+                max_answer_length=self.validation_config.max_answer_length,
+                version_2_with_negative=self.version_2_with_negative,
+                null_score_diff_threshold=self.validation_config.null_score_diff_threshold,
+                do_lower_case=self.do_lower_case,
+            )
+
+        logging.info(f"exact match {exact_match}")
+        logging.info(f"f1 {f1}")
+        tensorboard_logs = {'val_loss': avg_loss, 'exact_match': exact_match, 'f1': f1}
+        return {'val_loss': avg_loss, 'log': tensorboard_logs}
 
     def setup_training_data(self, train_data_config: Optional[DictConfig]):
         self._train_dl = self._setup_dataloader_from_config(cfg=train_data_config)
 
     def setup_validation_data(self, val_data_config: Optional[DictConfig]):
         self._validation_dl = self._setup_dataloader_from_config(cfg=val_data_config)
+        self.validation_config = val_data_config
 
     def setup_test_data(self, test_data_config: Optional[DictConfig]):
         self._test_dl = self._setup_dataloader(cfg=test_data_config)
```

Comment:
should this also be setup_dl_from_config?

Reply:
fixed